changefeedccl: only checkpoint leading spans above lead threshold

andyyang890 opened the pull request:
See individual commits

---

Fixes #137857

andyyang890 committed:
changefeedccl: only checkpoint leading spans above lead threshold

This patch updates the span-level checkpointing code to only checkpoint
spans that are above the lead threshold (formerly lag threshold) so that
we can at least maintain some minimal level of progress for spans that
we go through the effort of checkpointing.

Longer explanation: We create span-level checkpoints (at the configured
interval) if we are either in a backfill or there's at least one span
that is more than the lag/lead threshold ahead of the frontier min ts.
Before this change, in the latter case, we would checkpoint all spans
that are above the frontier min ts regardless of whether they are at
least the lag/lead threshold ahead. This change makes it so that when
we are creating a span-level checkpoint due to the lag/lead threshold,
we will only include spans that are ahead of the frontier min ts by at
least the lag/lead threshold so that we can maintain some minimal level
of progress for the spans that we checkpoint.

Release note: None

andyyang890 committed:
changefeedccl: rename changefeed.frontier_checkpoint_max_bytes setting

This patch renames `changefeed.frontier_checkpoint_max_bytes` to
`changefeed.span_checkpoint.max_bytes` for consistency.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138790)
<!-- Reviewable:end -->

andyyang890 committed:
changefeedccl: rename changefeed.frontier_checkpoint_frequency setting

This patch renames `changefeed.frontier_checkpoint_frequency` to
`changefeed.span_checkpoint.interval` for consistency.

Release note: None

stevendanna reviewed (commented):
I haven't looked closely so take this with a grain of salt, but I think this is the kind of change were we would benefit from the "painful elaboration of the obvious" in the commit message so that in a few months anyone partially familiar with changefeeds can understand we expect this change to do.

asg0451 commented on pkg/ccl/changefeedccl/resolvedspan/frontier.go:
```diff
@@ -170,7 +170,9 @@ func (f *CoordinatorFrontier) InBackfill(r jobspb.ResolvedSpan) bool {
 // MakeCheckpoint creates a checkpoint based on the current state of the frontier.
-func (f *CoordinatorFrontier) MakeCheckpoint(maxBytes int64) jobspb.ChangefeedProgress_Checkpoint {
+func (f *CoordinatorFrontier) MakeCheckpoint(
+	baselineTS hlc.Timestamp, maxBytes int64,
```
unused arg?

andyyang890 replied:
Good catch, fixed.

andyyang890 commented:
> I haven't looked closely so take this with a grain of salt, but I think this is the kind of change were we would benefit from the "painful elaboration of the obvious" in the commit message so that in a few months anyone partially familiar with changefeeds can understand we expect this change to do.

Thanks for the feedback, added a longer explanation in the commit message.

andyyang890 committed:
changefeedccl: rename changefeed.frontier_highwater_lag_checkpoint_threshold setting

This patch renames
`changefeed.frontier_highwater_lag_checkpoint_threshold`
to `changefeed.span_checkpoint.lead_threshold` for consistency.

Release note (ops change): The cluster setting
`changefeed.frontier_highwater_lag_checkpoint_threshold` has been
renamed to `changefeed.span_checkpoint.lead_threshold`. The old name
remains available for backwards-compatibility.

andyyang890 commented:
We discussed this change offline and decided the risks associated with this change aren't worth it given the planned work to improve span-level checkpoints. I pulled the first 3 commits renaming settings into a separate PR (#139064) and I'll close this PR soon.
